CARDS-1784: New question type: ResourceQuestion

marta- opened the pull request:
**Includes**:
* **[CARDS-1784](https://phenotips.atlassian.net/browse/CARDS-1784)** - New question type: ResourceQuestion
* Refactorinng: `VocabularyQuery` (the `VocabularyQuestion`'s suggested input) now leverages `ResourceQuery` (the `ResourceQuestion`'s suggested input)
* Misc bug fix: for VocabularyQuestions, the AnswerOption labels are  not added when previewing the questionnaire

**Testing**
* A `Resource Test` questionnaire was added to the `--test` mode
* Branch `CARDS-1737_CARDS-1784` (PR #1044 ) uses `ResourceQuestion` for the `Visit information/clinic` question implemented for CARDS-1737 (PR #1034).

**Reviewing**
* _Commit by commit_ is recommended, for a clearer view on how `ResourceQuery` (the suggested input for `ResourceQuestion`) emerged from `VocabularyQuery`.

marta- committed:
CARDS-1784: ResourceQuestion
Add a very basic implementation of the new question type allowing to choose as answers the paths of resources of a specified type that exist in the repo.

marta- committed:
CARDS-1784: ResourceQuestion
* Add two new display modes - input and list+input
* Implement suggested input ResourceQuery inspired by VocabularyQuery
* Reorganize VocabularyQuery styles to pull out reusable code between ResourceQuery and VocabularyQuery
* Add a label processor that fills in label for AnswerOptions to the ResourceQuestion's list+input displayMode

marta- committed:
CARDS-1784: ResourceQuestion
* Refactor the `Resource*LabelProcessor`s to reduce code duplication

marta- committed:
[misc] Bug fix: for VocabularyQuestions, the AnswerOption labels are not added when previewing the questionnaire

marta- committed:
CARDS-1784: ResourceQuestion
Remove unused `error` text

marta- committed:
CARDS-1784: ResourceQuestion
Refactored the handling of failed queries

marta- committed:
CARDS-1784: ResourceQuestion
Added Resource Test questionnaire and included the `Visit` subject in the --test mode

marta- committed:
CARDS-1784: ResourceQuestion
Order resource list by cards:defaultOrder (if present) or by the configured label property

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/labels/AbstractResourceLabelProcessor.java:
```diff
@@ -0,0 +1,91 @@
+import io.uhndata.cards.serialize.spi.ResourceJsonProcessor;
+
+/**
+ * Fills in the clinic answer for a given visit information clinic node.
```
It does not.

marta- replied:
Fixed in 97c0e49 .

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/labels/AbstractResourceLabelProcessor.java:
```diff
@@ -0,0 +1,91 @@
+    {
+        try {
+            if (question != null
+                && question.getProperty(PROP_RESOURCE_LABEL) != null
```
`getProperty` throws an exception if the property is not there. Checking `!=null` doesn't make sense, either the call succeeds and it is not null, or it throws and goes to the catch block.

```suggestion
                && question.hasProperty(PROP_RESOURCE_LABEL)
```

marta- replied:
Fixed in 97c0e49 .

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/labels/AbstractResourceLabelProcessor.java:
```diff
@@ -0,0 +1,91 @@
+                final Node resourceNode = resource.adaptTo(Node.class);
+                // Default the label to the resource name if the value for labelProperty is missing
+                if (labelPropertyName != null
+                    && resourceNode.getProperty(labelPropertyName) != null
```
```suggestion
                    && resourceNode.hasProperty(labelPropertyName)
```

marta- replied:
Fixed in 97c0e49 .

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/labels/ResourceLabelProcessor.java:
```diff
@@ -0,0 +1,99 @@
+import io.uhndata.cards.serialize.spi.ResourceJsonProcessor;
+
+/**
+ * Fills in the clinic answer for a given visit information clinic node.
```
Nope.

marta- replied:
Fixed in 97c0e49 .

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/labels/ResourceLabelProcessor.java:
```diff
@@ -0,0 +1,99 @@
+@Component(immediate = true)
+public class ResourceLabelProcessor extends AbstractResourceLabelProcessor implements ResourceJsonProcessor
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResourceLabelProcessor.class);
```
Unused, can be removed.

marta- replied:
Done in 97c0e49 .

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/labels/ResourceOptionsLabelProcessor.java:
```diff
@@ -0,0 +1,124 @@
+            Map<String, String> valueLabelMap = new LinkedHashMap<>();
+
+            Property valueProp = node.getProperty(PROP_VALUE);
+            if (valueProp.isMultiple()) {
```
...are answer options supposed to be multivalued? I don't think they are, so the whole method needs to extract one label only.

marta- replied:
There's another processor for answer option labels that assumes they could be multivalued. Indeed at this time they aren't multivalued, but the assumption doesn't break anything and we could possibly have this scenario in the future.

If you feel strongly about this, please make a task to refactor all label processors.

sdumitriu replied:
OK, I can see a possible use for multivalued answer options, so let's leave it as is.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/labels/ResourceOptionsLabelProcessor.java:
```diff
@@ -0,0 +1,124 @@
+@Component(immediate = true)
+public class ResourceOptionsLabelProcessor extends AbstractResourceLabelProcessor implements ResourceJsonProcessor
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResourceOptionsLabelProcessor.class);
```
Unused.

marta- replied:
Removed in 97c0e49 .

sdumitriu commented on modules/data-entry/src/main/frontend/src/resourceQuery/ResourceQuery.jsx:
```diff
@@ -0,0 +1,525 @@
+  const { clearOnClick, onClick, focusAfterSelecting, disabled, variant, isNested, placeholder,
+    value, questionDefinition, onChange, enableSelection, initialSelection, onRemoveOption, classes } = props;
+  const { maxAnswers, primaryType, labelProperty, propertiesToSearch } = questionDefinition;
+  const { enableUserEntry } = props;
```
Why isn't this two lines above?

marta- replied:
I find the code easier to read when we unpack props in groups that serve a specific purpose.

If you feel strongly about this feel free to put them all together before merging.

sdumitriu replied:
OK, I agree, I just thought we had a rule about importing things all together.

sdumitriu commented on modules/data-entry/src/main/frontend/src/resourceQuery/ResourceQuery.jsx:
```diff
@@ -0,0 +1,525 @@
+    value, questionDefinition, onChange, enableSelection, initialSelection, onRemoveOption, classes } = props;
+  const { maxAnswers, primaryType, labelProperty, propertiesToSearch } = questionDefinition;
+  const { enableUserEntry } = props;
+  const { fetchSuggestions, formatSuggestionData, infoDisplayer } = props;
```
Why aren't these three lines above?

marta- replied:
See previous comment.

sdumitriu commented on modules/data-entry/src/main/frontend/src/resourceQuery/ResourceQuery.jsx:
```diff
@@ -0,0 +1,525 @@
+      .catch(onFailure);
+  });
+
+  let shapeSuggestionData = (rawData, query) => {
```
I don't get this method name.

marta- replied:
See previous comment.

marta- replied:
Renamed in 61583fc .

sdumitriu reviewed (commented):
Huh. One unintended consequence of #989 is that the interpretations are returned for "question" suggests (e.g. `/Proms/Cardio/Step2/summary_score`).

marta- commented:
Demoting this PR to DRAFT for now, because I discovered a couple of filtering issues:
* regression: in the chips of vocabulary filters, the value is displayed (e.g. '/Vocabularies/HANCESTRO/HANCESTRO_####`) instead of the human-readable label
* missing feature: the form controls for the value of resource question filters are plain inputs instead of providing access to possible values

marta- committed:
CARDS-1784: ResourceQuestion
Addressing code review: comments and PropTypes improvements

marta- committed:
CARDS-1784: ResourceQuestion
Addressing code review: use state setter for buttonRef, proper use of `classnames` aggregator

Co-authored by: @OrdiNeu

marta- committed:
CARDS-1784: ResourceQuestion
Addressing review comments:
* Fixed inaccurate javadoc
* Fixed how we check that a node property exists
* Removed unused logger

Co-authored by: @sdumitriu

marta- committed:
CARDS-1784: ResourceQuestion
Added a resource filter allowing to filter by resource question answers on the dashboard

marta- commented:
> Huh. One unintended consequence of #989 is that the interpretations are returned for "question" suggests (e.g. `/Proms/Cardio/Step2/summary_score`).

Do you consider this a show stopper?

marta- committed:
CARDS-1784: ResourceQuestion
Addressing review comments: renamed method to something eaier to understand

sdumitriu commented:
> > Huh. One unintended consequence of #989 is that the interpretations are returned for "question" suggests (e.g. `/Proms/Cardio/Step2/summary_score`).
> 
> Do you consider this a show stopper?

Not at all, and definitely not for this PR. Do we have a better alternative for survey summaries?

marta- commented:
> * regression: in the chips of vocabulary filters, the value is displayed (e.g. '/Vocabularies/HANCESTRO/HANCESTRO_####`) instead of the human-readable label

Can no longer replicate, I may have been wrong.

> * missing feature: the form controls for the value of resource question filters are plain inputs instead of providing access to possible values

ResourceFilter added in 9d15217.

marta- committed:
CARDS-1784: ResourceQuestion
Addressing comments:
* Removed confusing links from the info card at the end of the Resource Test questionnaire
* Added descriptions to the vocab questions in Multiple Choice Test to specify the vocabulary filters configured for each question

sashaandjic reviewed (approved):
tested

Note: Concept Recognition Test form is not working properly for me (could be due to my environment as it works for Marta)

marta- commented:
Ready to merge in theory, but we should merge #1045 first and then rebase, re-review and re-test this one.

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/FilterComponents/ResourceFilter.jsx:
```diff
@@ -0,0 +1,77 @@
+const COMPARATORS = DEFAULT_COMPARATORS.slice().concat(UNARY_COMPARATORS);
+
+/**
+ * Display a filter on a vocabulary answer of a form. This is not meant to be instantiated directly, but is returned from FilterComponentManager's
```
```suggestion
 * Display a filter on a resource answer of a form. This is not meant to be instantiated directly, but is returned from FilterComponentManager's
```

marta- replied:
Fixed in 8d01174.

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/FilterComponents/ResourceFilter.jsx:
```diff
@@ -0,0 +1,77 @@
+ * Display a filter on a vocabulary answer of a form. This is not meant to be instantiated directly, but is returned from FilterComponentManager's
+ * getFilterComparatorsAndComponent method.
+ *
+ * @param {string} defaultValue The default value to place in the vocabulary filter
```
```suggestion
 * @param {string} defaultValue The default value to place in the filter
```

marta- replied:
Fixed in 8d01174.

marta- committed:
CARDS-1784: ResourceQuestion
Updating new components to mui 5+

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/FilterComponents/ResourceFilter.jsx:
```diff
@@ -0,0 +1,77 @@
+ * @param {string} defaultValue The default value to place in the vocabulary filter
+ * @param {func} onChangeInput Callback for when the value select has changed
+ * @param {object} questionDefinition Object containing the definition of the question. Should include "sourceVocabularies" and "vocabularyFilters" children.
+ * Other props are forwarded to the VocabularyQuery component
```
Needs updating.

marta- replied:
Updated in 8d01174.

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/FilterComponents/ResourceFilter.jsx:
```diff
@@ -0,0 +1,77 @@
+  defaultValue: PropTypes.string,
+  onChangeInput: PropTypes.func,
+  questionDefinition: PropTypes.shape({
+    sourceVocabularies: PropTypes.array,
```
Needs updating.

marta- replied:
Updated in 8d01174.

marta- commented:
Rebased on the latest `dev` which includes Material UI 5 upgrade, and updated the new components to Material UI 5 in 5af2f74.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ResourceQuestion.jsx:
```diff
@@ -0,0 +1,126 @@
+  // the list or dropdown.
+  // If the real number of resources matching the criteria specified by the question definition
+  // excedes this limit, ignore the specified displayMode and use a suggested input instead.
+  const MAX_TO_DISPLAY = 50;
```
I wonder if this should be configurable.

marta- replied:
Can we make it configurable as a separate task?

sdumitriu commented on modules/data-entry/src/main/frontend/src/resourceQuery/ResourceQuery.jsx:
```diff
@@ -0,0 +1,556 @@
+              }}
+            >
+              { /* Cover up a bug that causes the label to overlap the value:
+                   if it has a displayed value and isn't focused, don't show the label
```
I wonder if this bug has been fixed upstream and the workaround is no longer needed.

marta- replied:
Indeed it seems ok as far as I can tell. Committed the change in 56f1df8. We can revert after if the bug pops up again.

marta- committed:
CARDS-1784: ResourceQuestion
Addressing review comments: updated copy/pasted comments and proptypes

marta- committed:
[misc] Remove bug cover-up for fom control label bug that doesn't appear to be valid anymore

sashaandjic reviewed (approved):
re-tested
